redis/lookup.go

```diff
@@ -24,83 +26,103 @@ type Lookup struct {
 	// seconds.
 	updateFrequency time.Duration
 
+	// updateTimeout is the timeout for each SRV lookup. Defaults to 90% of the
+	// update frequency.
+	updateTimeout time.Duration
+
 	// dnsAuthority is the single <hostname|IPv4|[IPv6]>:<port> of the DNS
 	// server to be used for SRV lookups. If the address contains a hostname it
 	// will be resolved via the system DNS. If the port is left unspecified it
 	// will default to '53'. If this field is left unspecified the system DNS
 	// will be used for resolution.
 	dnsAuthority string
 
-	ring   *redis.Ring
-	logger blog.Logger
+	resolver *net.Resolver
+	ring     *redis.Ring
+	logger   blog.Logger
 }
 
-// NewLookup returns a new Lookup helper.
-func NewLookup(srvLookups []cmd.ServiceDomain, dnsAuthority string, frequency time.Duration, ring *redis.Ring, logger blog.Logger) *Lookup {
-	if frequency == 0 {
+// NewLookup constructs and returns a new Lookup instance. An initial SRV lookup
+// is performed to populate the Redis ring shards. If this lookup fails or
+// otherwise results in an empty set of resolved shards, an error is returned.
+func NewLookup(srvLookups []cmd.ServiceDomain, dnsAuthority string, frequency time.Duration, ring *redis.Ring, logger blog.Logger) (*Lookup, error) {
+	var lookup = &Lookup{
```

Comment:
I'm not a big fan of the "construct a partial object, then fill in the rest of its fields in conditionals" structure. It feels like it makes it very easy to forget to populate certain fields in some codepaths (e.g. currently lookup.dnsAuthority retains its zero-value if the input dnsAuthority is the empty string -- this makes sense, but currently _looks_ like an oversight rather than a conscious choice) and it's hard to keep track of where certain values are set (e.g. it took me two reads to realize that the conditional on line 65 was controlling both the dnsAuthority and the resolver fields).

I'd prefer a structure like:
```go
if frequency <= 0 {
  frequency = 30 * time.Second
}

var resolver *net.Resolver
if dnsAuthority != "" {
  ...
  dnsAuthority = stuff
  resolver = stuff
}

lookup := &Lookup{
  ...
}
```

Also even if you do keep things in this order, this should say `lookup := &Lookup{` instead of using `var`.

Reply:
That's fair, I've gone ahead and rewritten these bits to make sure that all of the construction takes place at once.